File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/chat/bridge/ChatCommandParser.java
package crabcraft.net.crabUtilities.chat.bridge;

import java.util.Locale;
import java.util.Set;

final class ChatCommandParser {

    private static final Set<String> PRIVATE_ALIASES =
            Set.of("msg", "message", "tell", "whisper", "w", "dm");
    private static final Set<String> REPLY_ALIASES = Set.of("r", "reply");
    private static final Set<String> STAFF_ALIASES = Set.of("sc", "staffchat");

    private ChatCommandParser() {}

    enum Type {
        NONE,
        PRIVATE,
        REPLY,
        STAFF
    }

    record Parsed(Type type, String target, String message) {
        boolean recognised() {
            return type != Type.NONE;
        }

        boolean valid() {
            return message != null && !message.isBlank()
                    && (type != Type.PRIVATE || target != null);
        }
    }

    static Parsed parse(String input) {
        if (input == null || input.length() < 2 || input.charAt(0) != '/') {
            return new Parsed(Type.NONE, null, null);
        }

        int separator = firstWhitespace(input, 1);
        String rawLabel = input.substring(1, separator < 0 ? input.length() : separator);
        String label = normaliseLabel(rawLabel);
        if (label == null) {
            return new Parsed(Type.NONE, null, null);
        }

        Type type;
        if (PRIVATE_ALIASES.contains(label)) {
            type = Type.PRIVATE;
        } else if (REPLY_ALIASES.contains(label)) {
            type = Type.REPLY;
        } else if (STAFF_ALIASES.contains(label)) {
            type = Type.STAFF;
        } else {
            return new Parsed(Type.NONE, null, null);
        }

        if (separator < 0) {
            return new Parsed(type, null, null);
        }
        int argumentsStart = skipWhitespace(input, separator);
        if (argumentsStart >= input.length()) {
            return new Parsed(type, null, null);
        }

        if (type != Type.PRIVATE) {
            return new Parsed(type, null, input.substring(argumentsStart));
        }

        ParsedTarget parsedTarget = parseTarget(input, argumentsStart);
        if (parsedTarget == null) {
            return new Parsed(type, null, null);
        }
        int messageStart = skipWhitespace(input, parsedTarget.end());
        if (messageStart >= input.length()) {
            return new Parsed(type, parsedTarget.value(), null);
        }
        return new Parsed(type, parsedTarget.value(), input.substring(messageStart));
    }

    private static String normaliseLabel(String rawLabel) {
        String lower = rawLabel.toLowerCase(Locale.ROOT);
        int namespace = lower.indexOf(':');
        if (namespace < 0) return lower;
        if (!lower.substring(0, namespace).equals("crabutilities")) return null;
        return lower.substring(namespace + 1);
    }

    private static ParsedTarget parseTarget(String input, int start) {
        if (input.charAt(start) != '"') {
            int end = firstWhitespace(input, start);
            if (end < 0) end = input.length();
            String target = input.substring(start, end);
            return target.isEmpty() ? null : new ParsedTarget(target, end);
        }

        StringBuilder target = new StringBuilder();
        boolean escaped = false;
        for (int i = start + 1; i < input.length(); i++) {
            char character = input.charAt(i);
            if (escaped) {
                if (character != '"' && character != '\\') return null;
                target.append(character);
                escaped = false;
            } else if (character == '\\') {
                escaped = true;
            } else if (character == '"') {
                int end = i + 1;
                if (end < input.length() && !Character.isWhitespace(input.charAt(end))) {
                    return null;
                }
                return target.isEmpty() ? null : new ParsedTarget(target.toString(), end);
            } else {
                target.append(character);
            }
        }
        return null;
    }

    private static int firstWhitespace(String input, int start) {
        for (int i = start; i < input.length(); i++) {
            if (Character.isWhitespace(input.charAt(i))) return i;
        }
        return -1;
    }

    private static int skipWhitespace(String input, int start) {
        int index = start;
        while (index < input.length() && Character.isWhitespace(input.charAt(index))) {
            index++;
        }
        return index;
    }

    private record ParsedTarget(String value, int end) {}
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/chat/bridge/ChatCommandParser.kt
package crabcraft.net.crabUtilities.chat.bridge

import java.util.Locale

object ChatCommandParser {
    private val PRIVATE_ALIASES = setOf("msg", "message", "tell", "whisper", "w", "dm")
    private val REPLY_ALIASES = setOf("r", "reply")
    private val STAFF_ALIASES = setOf("sc", "staffchat")

    enum class Type { NONE, PRIVATE, REPLY, STAFF }

    data class Parsed(private val type: Type, private val target: String?, private val message: String?) {
        fun type(): Type = type
        fun target(): String? = target
        fun message(): String? = message
        fun recognised(): Boolean = type != Type.NONE
        fun valid(): Boolean = message != null && message.any { !Character.isWhitespace(it) } && (type != Type.PRIVATE || target != null)
    }

    @JvmStatic
    fun parse(input: String?): Parsed {
        if (input == null || input.length < 2 || input[0] != '/') return Parsed(Type.NONE, null, null)
        val separator = firstWhitespace(input, 1)
        val rawLabel = input.substring(1, if (separator < 0) input.length else separator)
        val label = normaliseLabel(rawLabel) ?: return Parsed(Type.NONE, null, null)
        val type = when (label) {
            in PRIVATE_ALIASES -> Type.PRIVATE
            in REPLY_ALIASES -> Type.REPLY
            in STAFF_ALIASES -> Type.STAFF
            else -> return Parsed(Type.NONE, null, null)
        }
        if (separator < 0) return Parsed(type, null, null)
        val argumentsStart = skipWhitespace(input, separator)
        if (argumentsStart >= input.length) return Parsed(type, null, null)
        if (type != Type.PRIVATE) return Parsed(type, null, input.substring(argumentsStart))
        val parsedTarget = parseTarget(input, argumentsStart) ?: return Parsed(type, null, null)
        val messageStart = skipWhitespace(input, parsedTarget.end())
        if (messageStart >= input.length) return Parsed(type, parsedTarget.value(), null)
        return Parsed(type, parsedTarget.value(), input.substring(messageStart))
    }

    private fun normaliseLabel(rawLabel: String): String? {
        val lower = rawLabel.lowercase(Locale.ROOT)
        val namespace = lower.indexOf(':')
        if (namespace < 0) return lower
        if (lower.substring(0, namespace) != "crabutilities") return null
        return lower.substring(namespace + 1)
    }

    private fun parseTarget(input: String, start: Int): ParsedTarget? {
        if (input[start] != '"') {
            var end = firstWhitespace(input, start)
            if (end < 0) end = input.length
            val target = input.substring(start, end)
            return if (target.isEmpty()) null else ParsedTarget(target, end)
        }
        val target = StringBuilder()
        var escaped = false
        for (i in start + 1 until input.length) {
            val character = input[i]
            if (escaped) {
                if (character != '"' && character != '\\') return null
                target.append(character)
                escaped = false
            } else if (character == '\\') {
                escaped = true
            } else if (character == '"') {
                val end = i + 1
                if (end < input.length && !Character.isWhitespace(input[end])) return null
                return if (target.isEmpty()) null else ParsedTarget(target.toString(), end)
            } else {
                target.append(character)
            }
        }
        return null
    }

    private fun firstWhitespace(input: String, start: Int): Int {
        for (i in start until input.length) {
            if (Character.isWhitespace(input[i])) return i
        }
        return -1
    }

    private fun skipWhitespace(input: String, start: Int): Int {
        var index = start
        while (index < input.length && Character.isWhitespace(input[index])) index++
        return index
    }

    private data class ParsedTarget(private val value: String, private val end: Int) {
        fun value(): String = value
        fun end(): Int = end
    }
}
